Add (abstract) methods for global degrees to Python Graph

l-hoang opened the pull request:
This commit adds global degree methods to the Python graph object. For a
single host, global degree is equivalent to normal degree, but for
distributed execution, global degree allows a user to get the degree of
a node even if the host does not own all of the edges.

The overarching design goal behind this commit is that views that add on
to the base graph (like a bidirectional view) will be generated on
demand whenever a function is called (like global degree) which requires
a view to be built.

ddn0 commented:
The truest definition of a friend: one who knows where you hide.

roshandathathri commented on libgluon/src/Partition.cpp:
```diff
@@ -3442,16 +3416,19 @@ class Partitioner {
         DoEdgeAssignment(pview, opt_.mirror_edges);
     if (!topology_result) {
-      GALOIS_LOG_FATAL("{}", topology_result.error());
+      GALOIS_LOG_FATAL("failed to assign edges: {}", topology_result.error());
```
I think you meant ERROR

tylershunt replied:
ah yep good catch.

roshandathathri reviewed (approved):
Great!

gurbinder533 commented on libgluon/include/galois/graphs/DistGraph.h:
```diff
@@ -331,8 +331,9 @@ PartitionGraph(const DistOptions& opt) {
-    Partitioner part(dist.get(), opt);
-    part.Partition();
+    if (auto res = Partition(dist.get(), opt); !res) {
+      return res.error();
```
Why not `GALOIS_LOG_ERROR` here?

roshandathathri replied:
Logging at the source of the error is preferable typically

tylershunt replied:
Yeah there's no great discipline  about where/when we log errors yet. In this case I tried to log every case inside `Partition` and avoided double notification here.

gurbinder533 reviewed (approved):
LGTM.

Commit:
Add (abstract) methods for global degrees to Python Graph

This commit adds global degree methods to the Python graph object. For a
single host, global degree is equivalent to normal degree, but for
distributed execution, global degree allows a user to get the degree of
a node even if the host does not own all of the edges.

The overarching design goal behind this commit is that views that add on
to the base graph (like a bidirectional view) will be generated on
demand whenever a function is called (like global degree) which requires
a view to be built.

arthurp commented on python/katana/local/_graph.pyx:
```diff
@@ -301,6 +303,14 @@ cdef class GraphBase:
         return [EntityType.make(manager, typeid) for typeid in types]
 
+    @abstractmethod
+    def global_degree(self, uint64_t node):
```
Shouldn't this be `out_degree`? It's not the total degree.

l-hoang replied:
Works for me.

arthurp reviewed (approved):
I would rather have that function called `global_out_degree`. 

LGTM.
